parcels/tools/id_generators.py

```diff
@@ -0,0 +1,1036 @@
+            self._total_ids += 1
+            self._used_ids += 1
+            return np.uint64(result)
+        else:
+            result = self.released_ids.pop(n-1)
+            if self._track_id_index:
+                self._map_id_totalindex[result] = self._total_ids
+            self._total_ids += 1
+            self._used_ids += 1
+            return np.uint64(result)
+
+    def releaseID(self, id):
+        """
+        releases an ID, either by just removing it from the ID list or by
+        recovering the ID, reusing it for the next 'getID(...)' call
+        :arg id: ID to be released (and possibly recovered)
+        """
+        if not self._recover_ids:
+            return
+        self.released_ids.append(id)
+        self._used_ids -= 1
+
+    def preGenerateIDs(self, high_value):
+        """
+        abstract function - pre-allocating a range of IDs from 0 up to :arg high_value.
+        :arg high_value: (u)int64 value of the highest pre-generated ID itself.
+        """
+        if len(self.released_ids) > 0:
+            self.released_ids.clear()
+        self.released_ids = [i for i in range(0, high_value)]
+        self.next_id = high_value
+
+    def permuteIDs(self):
+        """
+        abstract function - randomizes pre-generated IDs
+        """
+        n = len(self.released_ids)
+        indices = random.randint(0, n, 2*n)
+        for index in indices:
+            id = self.released_ids.pop(index)
+            self.released_ids.append(id)
+
+    def __len__(self):
+        """
+        :returns the number of generated IDs
+        """
+        return self.next_id
+
+    def get_length(self):
+        """
+        function forward to
+        :returns the number of generated IDs
+        """
+        return len(self)
+
+
+class SpatialIdGenerator(BaseIdGenerator):
+    """
+    Generates 64-bit IDs based on 3D spatial coordinates, aiming for an ID distribution so that spatially-close
+    items have also close-by IDs.
+    """
+    _lon_bins = 360
+    _lat_bins = 180
+    _depth_bins = 32768
+    _lonbounds = np.zeros(2, dtype=np.float32)
+    _latbounds = np.zeros(2, dtype=np.float32)
+    _depthbounds = np.zeros(2, dtype=np.float32)
+    local_ids = None
+    released_ids = {}
+
+    def __init__(self, lon_bins=360, lat_bins=180, depth_bins=32768):
+        """
+        SpatialIdGenerator - Constructor
+
+        ID generator that manages IDs in a spatial mapping scheme, so that IDs
+        that are spatially close are also numerically close.
+
+        Attention: the bins are used in a bit allocation scheme so that
+        (log2(arg:lon_bins) * long2(arg::lat_bins) * log2(arg:depth_bins)) <= 32
+        """
+        super(SpatialIdGenerator, self).__init__()
+        self._lonbounds = np.array([-180.0, 180.0], dtype=np.float32)
+        self._latbounds = np.array([-90.0, 90.0], dtype=np.float32)
+        self._depthbounds = np.array([0.0, 1.0], dtype=np.float32)
```

Comment:
Why not min and max depth here?

Reply:
This comment still needs to be addressed

Reply:
the information is not automatically available, thus we cannot get the min-max from the data, especially not on ID generation (because there are no data when the generator is created). It could be that you mean the min. and max. of float32 or float64, but that isn't beneficial: if the bounds are 10^-104 to 10^104, but the user only has depths from 0 to 1000m, then all coordinates would get the same depth bit allocation - which makes no sense if one actually uses a 3D hash. So, if one really wants to have a different depth range from the start than what is suggested and one doesn't like the later `setDepthLimits`, then I need a more specific suggestion on what those bounds should be to address it.